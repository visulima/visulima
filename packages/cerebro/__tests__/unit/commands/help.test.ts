import { beforeEach, describe, expect, it, vi } from "vitest";

import type { Command as ICommand } from "../../../src/@types/command";
import type { Toolbox as IToolbox } from "../../../src/@types/toolbox";
import HelpCommand from "../../../src/commands/help-command";
import globalOptions from "../../../src/default-options";
import commandLineUsage from "../../../src/util/command-line-usage";

vi.mock(import("../../../src/util/command-line-usage"));
vi.mock(import("../default-options"), () => {
    return {
        default: [
            // Your default options structure here
        ],
    };
});

const loggerMock = {
    debug: vi.fn(),
    raw: vi.fn(),
    warn: vi.fn(),
};

const runtimeMock = {
    getCliName: vi.fn().mockReturnValue("testcli"),
    getCommandSection: vi.fn().mockReturnValue({ footer: "Test footer", header: "Test header" }),
};

const headerSection = {
    content: "testcli <command> [positional arguments] [options]",
    header: " Usage ",
};

const footerSection = {
    content: "Run \"testcli help <command>\" or \"testcli <command> --help\" for help with a specific command.",
    raw: true,
};

const globalOptionsOptionsList = {
    header: " Global Options ",
    optionList: globalOptions,
};

describe("command/help", () => {
    const commandsMap = new Map<string, ICommand>([]);

    beforeEach(() => {
        vi.clearAllMocks();

        commandsMap.clear();

        commandsMap.set("test", {
            description: "A test command",
            execute: () => {},
            group: "test",
            name: "test",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });
    });

    it("should print general help", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });

    it("should print command-specific help", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "test",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });

    it("should print general help and not include hidden commands", () => {
        expect.assertions(4);

        const secretCommandsMap = new Map([
            // Add a hidden command for testing
            [
                "secret",
                {
                    execute: () => {},
                    hidden: true,
                    name: "secret",
                },
            ],
            [
                "test",
                {
                    description: "A test command",
                    execute: () => {},
                    name: "test",
                    options: [{ description: "Enable verbose output", name: "verbose" }],
                },
            ],
        ]);

        const helpCommand = new HelpCommand(secretCommandsMap);
        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");

        // Ensure hidden commands are not included in the output
        const usageCalls = vi.mocked(commandLineUsage).mock.calls[0][0];

        expect(usageCalls.some((section) => section.content?.includes("secret"))).toBe(false);
    });

    it("should display aliases if present", () => {
        expect.assertions(3);

        const commandWithAlias = {
            alias: ["cwa", "aliasForCommand"],
            execute: () => {},
            name: "commandWithAlias",
        };

        commandsMap.set(commandWithAlias.name, commandWithAlias);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "commandWithAlias",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });

    it("should display command examples", () => {
        expect.assertions(3);

        const commandWithExamples = {
            examples: ["testcli test --verbose", "testcli test --verbose --debug"],
            execute: () => {},
            name: "commandWithExamples",
        };

        commandsMap.set(commandWithExamples.name, commandWithExamples);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "commandWithExamples",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });

    it("should display a empty help command if no commands exists", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(new Map());

        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });

    it("should display group of commands if the group option is passed", () => {
        expect.assertions(3);

        commandsMap.set("test2", {
            description: "A test2 command",
            execute: () => {},
            group: "test",
            name: "test2",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });

        commandsMap.set("no-group-test", {
            description: "A test command",
            execute: () => {},
            name: "no-group-test",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });

        const helpCommand = new HelpCommand(commandsMap);

        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            options: {
                group: "test",
            },
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenNthCalledWith(1, "Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenNthCalledWith(3, "Test footer");
    });
});
