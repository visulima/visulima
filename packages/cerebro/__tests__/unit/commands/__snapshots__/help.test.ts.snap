// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`command/help > should display a empty help command if no commands exists 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32m<command>[39m [positional arguments] [33m[options][39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "[36mtestcli[39m [32mhelp <command>[39m" or "[36mtestcli[39m [32m<command>[39m [33m--help[39m" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should display aliases if present 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32mcommandWithAlias[39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "content": [
            "cwa",
            "aliasForCommand",
          ],
          "header": "Alias(es)",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should display command examples 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32mcommandWithExamples[39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": [
            "testcli test --verbose",
            "testcli test --verbose --debug",
          ],
          "header": "Examples",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should display group of commands if the group option is passed 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32m<command>[39m [positional arguments] [33m[options][39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "content": [
            [
              "[32mtest[39m ",
              "A test command",
            ],
            [
              "[32mtest2[39m ",
              "A test2 command",
            ],
          ],
          "header": "[7m[32m Available Test Commands [39m[27m",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "[36mtestcli[39m [32mhelp <command>[39m" or "[36mtestcli[39m [32m<command>[39m [33m--help[39m" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should print command-specific help 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32mtest[39m [options]",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "content": "A test command",
          "header": "[7m[32m Description [39m[27m",
        },
        {
          "header": "[7m[33m Command Options [39m[27m",
          "optionList": [
            {
              "description": "Enable verbose output",
              "name": "verbose",
            },
          ],
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should print general help 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32m<command>[39m [positional arguments] [33m[options][39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "content": [
            [
              "[32mtest[39m ",
              "A test command",
            ],
          ],
          "header": " [7m[32m Test [39m[27m",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "[36mtestcli[39m [32mhelp <command>[39m" or "[36mtestcli[39m [32m<command>[39m [33m--help[39m" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should print general help and not include hidden commands 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        {
          "content": "[36mtestcli[39m [32m<command>[39m [positional arguments] [33m[options][39m",
          "header": "[7m[36m Usage [39m[27m",
        },
        {
          "content": [
            [
              "[32mtest[39m ",
              "A test command",
            ],
          ],
          "header": "[7m[32m Available Commands [39m[27m",
        },
        {
          "header": "[7m[33m Global Options [39m[27m",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "[36mtestcli[39m [32mhelp <command>[39m" or "[36mtestcli[39m [32m<command>[39m [33m--help[39m" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
